### ISSUE TITLE
Hikari and Tomcat 

### ISSUE THREAD

**Issue by Eithcowich** [brettwooldridge/HikariCP#836]
#### Environment
```
HikariCP version: 2.6.0
JDK version     : 1.8.0_121
Database        :  10.0.14-MariaDB-log
Driver version  : mysql-connector-java-5.0.8-bin.jar
```
-----------------------------------------------------------------------------------------

Hi,

I'm using HikariCP with a Tomcat Web application against MySql. Every time I call my web app Hikari opens 5 new connections (the max we set.) This is obviously not what we want. How do we fix it, so that connections are reused? 

I have the Hikari jar inside tomcat/lib. When it was only in the app we got a ClassNotFound exception. 

The code I'm using: 

				dataSource.setDriverClassName("com.mysql.jdbc.Driver");
				dataSource.setJdbcUrl(dbURL+ currentDB);
				dataSource.setUsername(user);
				dataSource.setPassword(passwd);

				dataSource.setMaximumPoolSize(5);
				dataSource.setMinimumIdle(5);

				dataSource.setLeakDetectionThreshold(15000);
				dataSource.setConnectionTimeout(50000);

				dataSource.setConnectionTestQuery("SELECT 1");

				sessionConnection = dataSource.getConnection();
				sessionStatement = sessionConnection.createStatement();

**Comment by billoneil** [comment]
It sounds like you are creating a new datasource on every request instead
of reusing a singleton. How exactly is the datasource being instantiated?

Bill

**Comment by Eithcowich** [comment]
It's a class variable, 	private final HikariDataSource dataSource = new HikariDataSource();

**Comment by billoneil** [comment]
Is that class getting constructed on every request? It sounds like it is.
You can either make the datasource static or pass it into the constructor /
inject it.

**Comment by Eithcowich** [comment]
You're right, it does create a new datasource every time. Thanks, I'll fix it.

**Comment by Eithcowich** [comment]
So, actually, I don't need to configure tomcat the way it says here? https://github.com/brettwooldridge/HikariCP/wiki/JNDI-DataSource-Factory-(Tomcat,-etc.)

I didn't, I just put the jar in tomcat/lib

**Comment by billoneil** [comment]
You can use that if you are using JNDI. You can also just create a new
HikariDataSource yourself.

**Comment by Eithcowich** [comment]
So far so good. But I'm seeing that killed processes are still around in my process list. This is probably not related to rollback issues in the db because the queries where very simple, and all of them were selects. So why are the processes still around? 

And thank you for continuing to assist with this.

**Comment by billoneil** [comment]
What do you mean by killed processes? Java processes? Or open db
connections?

The connection pool is supposed to reuse connections. According to your
config you should always have 5 connections open to the database. If you
run a query the connection is returned to the pool instead of being closed.
If you forcibly kill a connection the pool should open a new one since you
have minimum idle = 5.

**Comment by Eithcowich** [comment]
Yes, I understand how the pool should work. Here's a screenshot. I currently have 24 db processes, 9 open and 15 "killed". The earlier killed ones are around for over 30 minutes now. I'm not sure why would that happen.

![screen shot 2017-03-07 at 3 21 41 pm](https://cloud.githubusercontent.com/assets/6975895/23658093/de7e453c-0349-11e7-8612-772204fa757d.png)

**Comment by Eithcowich** [comment]
It looks like after several minutes of inactivity the connections are killed - and stay in the process list as "killed" - and the next time I call my app I get a nullPointerException trying to call a query on a connection - and the pool starts opening the connections again. 
Opening 5 connections takes several seconds. 
Does any of this makes sense?

**Comment by brettwooldridge** [comment]
@Eithcowich ``dataSource.setMinimumIdle(5)`` is going to guarantee that there are always 5 connections available.  If you want your pool to shrink when not in use, you need to set this to something lower than your max. connections.

**Comment by brettwooldridge** [comment]
@Eithcowich Also, if you enable DEBUG level logging (in log4j or whatever you are using) for package ``com.zaxxer.hikari`` you should see when HikariCP is closing connections, adding connections to the pool, etc.
